### ISSUE TITLE
Proposal: Allow methods to have "static" variables.

### ISSUE THREAD

**Issue by TheUnlocked** [dotnet/csharplang#1247]
Often, one will want to create a private field for the purpose of having a persistent value in a method. Consider this example:

```cs
private int nextNumIter = 0;
public void PrintNextNum()
{
     Console.WriteLine(nextNumIter++);
}
```

The problem with this is that it may break encapsulation. `nextNumIter` should _only_ be used for `NextNum`, and no other methods in the class should be able to access it. As written however, that isn't true.

 The proposal is to add a more encapsulated (and, in my opinion, more elegant) syntax for persistent variables in methods by allowing the use of static, like so:

```cs
public void NextNum()
{
     static int iter = 0;
     Console.WriteLine(iter++);
}
```

The following rules and restrictions would likely apply:

- Static method variables may _only_ be defined at the start of a function.
- Static method variables are associated the "instance," even though they're placed inside of a function.
     - A static method variable inside of a non-static method would act as a hidden non-static field.
     - A static method variable inside of a static method would act as a hidden static field.

There are of course some drawbacks:

- Potentially confusing. A keyword other than `static` might be clearer?
- This is already possible, the proposal would just add syntactic sugar and extra encapsulation.

**Comment by HaloFour** [comment]
#832

**Comment by MkazemAkhgary** [comment]
>The problem with this is that it may break encapsulation.

private field is already encapsulatet by enclosing class. The way you treat private field relies on implementation of that class. From outside, it doesnt matter since its implemetation detail.

>nextNumIter should only be used for NextNum, and no other methods in the class should be able to access it

Then dont use it. If you see that problematic, break your class into smaller pieces.

**Comment by Neme12** [comment]
"Static" variable associated with an instance? No - we don't need `static` to have 10 different meanings like it does in C++. Also, what if you wanted an actual `static` variable in a non-static method?

**Comment by jnm2** [comment]
Please close and move the conversation to the existing #832.

**Comment by TheUnlocked** [comment]
Issue being closed, as #832 is already acting as a discussion area for this feature request.
